File: combinatorics/resource/ST/project_5/bai_toan_4/mutil_graph/al_to_amap.cpp
#include <iostream>
#include <vector>
#include <map>
#include <sstream>

using namespace std;

// Ham chuyen AL sang AMap
map<int, map<int, int>> alToAmap(const vector<vector<int>>& adjList) {
    map<int, map<int, int>> amap;

    for (int u = 0; u < adjList.size(); ++u) {
        for (int v : adjList[u]) {
            amap[u][v]++;  // Tang dem moi khi co canh (u,v)
        }
    }

    return amap;
}

int main() {
    int n;
    cout << "Nhap so dinh: ";
    cin >> n;

    vector<vector<int>> adjList(n);
    cin.ignore(); // Xoa bo dem

    cout << "Nhap danh sach ke:\n";
    for (int i = 0; i < n; ++i) {
        cout << "Ke cua dinh " << i << ": ";
        string line;
        getline(cin, line);
        istringstream iss(line);
        int x;
        while (iss >> x)
            adjList[i].push_back(x);
    }

    // Chuyen doi va in ket qua
    auto amap = alToAmap(adjList);
    cout << "Ban do ke (AMap):\n";
    for (const auto& [u, neighbors] : amap) {
        cout << u << ": ";
        for (const auto& [v, w] : neighbors) {
            cout << "(" << v << "," << w << ") ";
        }
        cout << endl;
    }

    return 0;
}

File: combinatorics/resource/ST/project_5/bai_toan_4/mutil_graph/al_to_amap.py
from collections import defaultdict

def al_to_amap(adj_list):
    amap = defaultdict(lambda: defaultdict(int))  # Khoi tao dict 2 chieu

    for u, neighbors in enumerate(adj_list):
        for v in neighbors:
            amap[u][v] += 1  # Tang so lan xuat hien cua canh (u,v)

    return amap

# Nhap tu ban phim
n = int(input("Nhap so dinh: "))
adj_list = []

print("Nhap danh sach ke:")
for i in range(n):
    row = list(map(int, input(f"Ke cua dinh {i}: ").split()))
    adj_list.append(row)

# Chuyen doi va in ket qua
amap = al_to_amap(adj_list)

print("Ban do ke (AMap):")
for u in amap:
    print(f"{u}: ", end="")
    for v, w in amap[u].items():
        print(f"({v},{w})", end=" ")
    print()